server/sonar-server/src/main/java/org/sonar/server/computation/step/FeedPeriodsStep.java

```diff
@@ -0,0 +1,278 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.step;
+
+import com.google.common.base.Strings;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.List;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.config.Settings;
+import org.sonar.api.resources.Qualifiers;
+import org.sonar.api.utils.DateUtils;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonar.batch.protocol.output.BatchReport;
+import org.sonar.core.component.ComponentDto;
+import org.sonar.core.component.SnapshotDto;
+import org.sonar.core.component.SnapshotQuery;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.server.computation.batch.BatchReportReader;
+import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.component.TreeRootHolder;
+import org.sonar.server.computation.period.Period;
+import org.sonar.server.computation.period.PeriodsHolderImpl;
+import org.sonar.server.db.DbClient;
+
+import static org.sonar.core.component.SnapshotQuery.SORT_FIELD.BY_DATE;
+import static org.sonar.core.component.SnapshotQuery.SORT_ORDER.ASC;
+import static org.sonar.core.component.SnapshotQuery.SORT_ORDER.DESC;
+
+/**
+ * Populates the {@link org.sonar.server.computation.period.PeriodsHolder}
+ *
+ * Here is how these periods are computed :
+ * - Read the 5 period properties ${@link CoreProperties#TIMEMACHINE_PERIOD_PREFIX}
+ * - Try to find the matching snapshots from the properties
+ * - If a snapshot is found, a new period is added to the repository
+ */
+public class FeedPeriodsStep implements ComputationStep {
+
+  private static final Logger LOG = Loggers.get(FeedPeriodsStep.class);
+
+  private static final int NUMBER_OF_PERIODS = 5;
+
+  private final DbClient dbClient;
+  private final Settings settings;
+  private final TreeRootHolder treeRootHolder;
+  private final BatchReportReader batchReportReader;
+  private final PeriodsHolderImpl periodsHolder;
+
+  public FeedPeriodsStep(DbClient dbClient, Settings settings, TreeRootHolder treeRootHolder, BatchReportReader batchReportReader,
+    PeriodsHolderImpl periodsHolder) {
+    this.dbClient = dbClient;
+    this.settings = settings;
+    this.treeRootHolder = treeRootHolder;
+    this.batchReportReader = batchReportReader;
+    this.periodsHolder = periodsHolder;
+  }
+
+  @Override
+  public void execute() {
+    periodsHolder.setPeriods(createPeriods());
+  }
+
+  private List<Period> createPeriods() {
+    List<Period> periods = new ArrayList<>();
```

Comment:
this variable is defined out of its minimal scope

in addition, it might be initialized (ie. an array created and memory claimed) even if there is actually nothing added to it. Move this variable to the branch where projectDto is not `null` and return Collections.emptyList in other branches